xdsl/parser/attribute_parser.py

```diff
@@ -670,61 +672,174 @@ def _parse_optional_builtin_parametrized_attr(self) -> Attribute | None:
+                parsed_int = int.from_bytes(
+                    byte_list[i * chunk_size : (i + 1) * chunk_size],
+                    sys.byteorder,
+                    signed=True,
+                )
+                data_values.append(parsed_int)
+        if len(data_values) == 1:
+            # Splat attribute case, same value everywhere,
+            # Emit values repeatedly and emit empty shape
+            return [data_values[0]] * math.prod(type.get_shape()), []
+        return data_values, [num_chunks]
+
     def _parse_builtin_dense_attr(self, _name: Span) -> DenseIntOrFPElementsAttr:
-        self.parse_punctuation("<", " in dense attribute")
+        def _parse_dense_type():
+            # Parse the dense type and check for correctness
+            self.parse_punctuation(":", " in dense attribute")
+            type = self.expect(
+                self.parse_optional_type, "Dense attribute must be typed!"
+            )
+            # Check that the type is correct.
+            if not isa(
+                type,
+                RankedVectorOrTensorOf[IntegerType]
+                | RankedVectorOrTensorOf[IndexType]
+                | RankedVectorOrTensorOf[AnyFloat],
+            ):
+                self.raise_error(
+                    "Expected vector or tensor type of " "integer, index, or float type"
+                )
 
-        # The flatten list of elements
+            # Check for static shapes in type
+            if any(dim == -1 for dim in list(type.get_shape())):
+                self.raise_error("Dense literal attribute should have a static shape.")
+
+            return type
+
+        def _check_list_dense_attr(type_shape: list[int], shape: list[int] | None):
+            # Check that the shape matches the data when given a shaped data.
+            # For splat attributes any shape is fine
+            if shape == []:
+                return
+            if shape is not None and type_shape != shape:
+                self.raise_error(
+                    f"Shape mismatch in dense literal. Expected {type_shape} "
+                    f"shape from the type, but got {shape} shape."
+                )
+
+        def _check_hex_dense_attr(type_num_values: int, shape: list[int]):
+            # perform check on just the flattened shapes for hex attribute
+            assert shape is not None
+            # For splat attributes any shape is fine
+            if shape == []:
+                return
+            if type_num_values != shape[0]:
+                self.raise_error(
+                    f"Shape mismatch in dense literal. Expected {type_num_values} "
+                    f"elements from the type, but got {shape[0]} elements."
+                )
+
+        def _check_empty_dense_attr(type_num_values: int):
+            if type_num_values != 0:
+                self.raise_error(
+                    "Expected at least one element in the "
+                    "dense literal, but got None"
+                )
```

Comment:
What's the purpose of these helpers? I see that they factor logic but also obscure the flow a little, and make the diff slightly messier. Do you think they'll be called more than once in the future? My first instinct is to inline them again.

Reply:
I disagree though. The code before was really long and that makes things harder to read.
With these functions, it's easier to understand why this code works tbh.
The thing I would change though is maybe make a big if/else in the code for the hex case, since we can use the `parse_dense_type` can be reused

Reply:
But that's because the logic of the function is too convoluted. @JosseVanDelm, do you mind if I try to refactor it a little bit? I think it could benefit from an early exit in the empty case, and from encapsulating the hex parsing logic to a single section of code.

Reply:
I played with this locally, and it looks like my initial intuition about early exit was wrong, but I found another way to refactor this little function that I feel is beneficial, not sure how best to share the snippet for your consideration.

Reply:
@superlopuh let me try to add @math-fehr's suggestion first